### PR TITLE
whisper : add context param for disable gpu

### PR DESCRIPTION
**Pull request by jhen0409**
Currently the Metal backend is using some SIMD operations, which it's only supported for Apple7+ family devices (ref: [Metal-Feature-Set-Tables.pdf](https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf)). In order to allow most older devices to run whisper.cpp normally, we can provide a param like `use_gpu`.

I think it will be helpful for some whisper.cpp binding that enabled Metal in build, but need to support old devices, or just don't want to access the GPU resources in some cases.

In this PR, I've added a new `struct whisper_context_params { use_gpu = true }` for that, I think it will also used for param like `use_mmap` in the future.

@ggerganov please let me know if you think this is a good idea or not, if yes I will do update for other backends & all bindings.

### CODE REVIEW

**Comment by bobqianic** [comment]
It seems that this modification could introduce some backward compatibility issues, which would necessitate refactoring many of the APIs. I recall the last time I attempted to introduce a `debug-mode` option in whisper.cpp. It proved challenging to implement without disrupting the existing API. Generally speaking, I believe **_the current API design lacks elegance_**. A well-designed API should ensure both forward and backward compatibility, so that minor changes don't inconvenience or disrupt the users.

**Comment by jhen0409** [comment]
> It seems that this modification could introduce some backward compatibility issues, which would necessitate refactoring many of the APIs. I recall the last time I attempted to introduce a `debug-mode` option in whisper.cpp. It proved challenging to implement without disrupting the existing API. Generally speaking, I believe **_the current API design lacks elegance_**. A well-designed API should ensure both forward and backward compatibility, so that minor changes don't inconvenience or disrupt the users.

We can consider to add new methods and deprecate the old methods, it can retain some flexibility for users.

**Comment by jhen0409** [comment]
> It seems that this modification could introduce some backward compatibility issues, which would necessitate refactoring many of the APIs. I recall the last time I attempted to introduce a `debug-mode` option in whisper.cpp. It proved challenging to implement without disrupting the existing API. Generally speaking, I believe **_the current API design lacks elegance_**. A well-designed API should ensure both forward and backward compatibility, so that minor changes don't inconvenience or disrupt the users.

We can consider to add new methods and mark deprecated for the old methods, it can retain some flexibility for users.

(There were some issues on Comments with my phone that made me delete the previous comment, sorry for the double notifications.)

**Comment by bobqianic** [comment]
@ggerganov ping

**Comment by ggerganov** [comment]
Will take a look this week - been travelling for a few days

**Comment by ggerganov** [comment]
@jhen0409 Yes, this is a good idea

**Comment by jhen0409** [comment]
I'm looking for a way to disable CUDA backend (and OpenCL) by the param, should we just check `GGML_BACKEND_GPU` or need add a param to ggml?

I found all tensors are not setup backend as `GGML_BACKEND_GPU`, and it only use `ggml_cuda_mul_mat` (during can_mul_mat(...) will be true). Compared to llama.cpp, I think it missing setup such like a call of `ggml_cuda_assign_buffers`, but I'm not still not familiar with it. Also, it seems like related to #1179.

File: ggml-metal.m [hunk 1]
```diff
@@ -150,56 +150,46 @@ @implementation GGMLMetalClass
...
+        } else {
+            metal_printf("%s: default.metallib not found, loading from source\n", __func__);
+
+            NSString * sourcePath = [bundle pathForResource:@"ggml-metal" ofType:@"metal"];
+            metal_printf("%s: loading '%s'\n", __func__, [sourcePath UTF8String]);
+            NSString * src = [NSString stringWithContentsOfFile:sourcePath encoding:NSUTF8StringEncoding error:&error];
+            if (error) {
+                metal_printf("%s: error: %s\n", __func__, [[error description] UTF8String]);
+                return NULL;
+            }
```
Comment (jhen0409) [inline ggml-metal.m:178]: I enabled metal for test whisper.swiftui, and the project required to load compiled `default.metallib`, so I made this change. @bachittle I think this should be also help for https://github.com/ggerganov/llama.cpp/issues/3284.

For `GGML_SWIFT` I use `SWIFTPM_MODULE_BUNDLE` instead and reuse the code of default.metallib load, I think it should be also work in llama.cpp swift package.

**Comment by jhen0409** [comment]
Aside from the GPU backend questions, I think another things are ready to review.

**Comment by paulocoutinhox** [comment]
It solve this https://github.com/ggerganov/whisper.cpp/issues/1386?

**Comment by bobqianic** [comment]
> I found all tensors are not setup backend as GGML_BACKEND_GPU, and it only use ggml_cuda_mul_mat (during can_mul_mat(...) will be true). Compared to llama.cpp, I think it missing setup such like a call of ggml_cuda_assign_buffers, but I'm not still not familiar with it. Also, it seems like related to https://github.com/ggerganov/whisper.cpp/issues/1179.

I'm in a similar situation. I'm completely lost in the llama.cpp code. Can't figure out how to offload the other operations to the GPU.

**Comment by ggerganov** [comment]
> Aside from the GPU backend questions, I think another things are ready to review.

I will implement full GPU offloading in the following days.
Turning off CUDA is currently discussed in https://github.com/ggerganov/llama.cpp/pull/3946

**Review by ggerganov — APPROVED** [review]
@jhen0409 This should be good to merge, correct?

**Comment by jhen0409** [comment]
> @jhen0409 This should be good to merge, correct?

Yes, it should be ready.

**Pull request merged**